src/pke/lib/scheme/ckksrns/ckksrns-schemeswitching.cpp

```diff
@@ -0,0 +1,2152 @@
+                g_coefficientsFHEW128_8;  // If the output messages are bits, we could use a lower degree polynomial
+        }
+        else {
+            coefficientsFHEW = g_coefficientsFHEW128_9;
+        }
+    }
+
+    // Step 1. Form matrix A and vector b from the LWE ciphertexts, but only extract the first necessary number of them
+    std::vector<std::vector<std::complex<double>>> A(numValues);
+
+    // To have the same encoding as A*s, create b with the appropriate number of elements
+    uint32_t b_size = numValues;
+    if ((numValues % n) != 0) {
+        b_size = numValues + n - (numValues % n);
+    }
+    std::vector<std::complex<double>> b(b_size);
+
+    double prescale =
+        (1.0 / LWECiphertexts[0]->GetModulus().ConvertToDouble()) /
+        K;  // Combine the scale with the division by K to consume fewer levels, but careful since the value might be too small
+
+#pragma omp parallel for
+    for (uint32_t i = 0; i < numValues; i++) {
+        auto a = LWECiphertexts[i]->GetA();
+        A[i]   = std::vector<std::complex<double>>(a.GetLength());
+        for (uint32_t j = 0; j < a.GetLength(); j++) {
+            A[i][j] = std::complex<double>(a[j].ConvertToDouble(), 0);
+        }
+        b[i] = std::complex<double>(LWECiphertexts[i]->GetB().ConvertToDouble(), 0);
+    }
+
+    // Step 2. Perform the homomorphic linear transformation of A*skLWE
+    Ciphertext<DCRTPoly> AdotS = EvalPartialHomDecryption(*ccCKKS, A, m_FHEWtoCKKSswk, m_dim1FC, prescale, 0);
+
+    // Step 3. Get the ciphertext of B - A*s
+    for (uint32_t i = 0; i < numValues; i++) {
+        b[i] = b[i] * prescale;
+    }
+    Plaintext BPlain = ccCKKS->MakeCKKSPackedPlaintext(b, AdotS->GetNoiseScaleDeg(), AdotS->GetLevel(), nullptr, N / 2);
+
+    auto BminusAdotS = ccCKKS->EvalAdd(ccCKKS->EvalNegate(AdotS), BPlain);
+
+    if (cryptoParamsCKKS->GetScalingTechnique() == FIXEDMANUAL) {
+        ccCKKS->ModReduceInPlace(BminusAdotS);
+    }
+    else {
+        if (BminusAdotS->GetNoiseScaleDeg() == 2)
+            ccCKKS->GetScheme()->ModReduceInternalInPlace(BminusAdotS, BASE_NUM_LEVELS_TO_DROP);
+    }
+
+    // Step 4. Do the modulus reduction: homomorphically evaluate modular function. We do it by using sine approximation.
+
+    auto BminusAdotS2 = BminusAdotS;  // Instead of zeroing out slots which are not of interest as done above
+
+    double a_cheby = -1;
+    double b_cheby = 1;  // The division by K was performed before
+
+    // double a_cheby = -K; double b_cheby = K; // Alternatively, do this separately to not lose precision when scaling with everything at once
+    // auto BminusAdotS2 = BminusAdotS;
+
+    auto BminusAdotS3 = ccCKKS->EvalChebyshevSeries(BminusAdotS2, coefficientsFHEW, a_cheby, b_cheby);
+
+    if (cryptoParamsCKKS->GetScalingTechnique() != FIXEDMANUAL) {
+        ccCKKS->GetScheme()->ModReduceInternalInPlace(BminusAdotS3, BASE_NUM_LEVELS_TO_DROP);
+    }
+
+    int32_t r = 3;
+    for (int32_t j = 1; j < r + 1; j++) {
```

Comment:
size_t instead of int32_t

Reply:
j should be signed, changed r to an enum.